include/awkward/builder/ArrayBuilder.h

```diff
@@ -7,369 +8,329 @@
 #include <string>
 #include <vector>
 
-#include "awkward/common.h"
-#include "awkward/builder/Builder.h"
 #include "awkward/BuilderOptions.h"
+#include "awkward/builder/Builder.h"
+#include "awkward/common.h"
 
 namespace awkward {
-  class Builder;
-  using BuilderPtr = std::shared_ptr<Builder>;
+class Builder;
+using BuilderPtr = std::shared_ptr<Builder>;
+
+/// @class ArrayBuilder
+///
+/// @brief User interface to the Builder system: the ArrayBuilder is a
+/// fixed reference while the Builder subclass instances change in
+/// response to accumulating data.
+class LIBAWKWARD_EXPORT_SYMBOL ArrayBuilder {
+public:
+  /// @brief Creates an ArrayBuilder from a full set of parameters.
+  ///
+  /// @param options Configuration options for building an array;
+  /// these are passed to every Builder's constructor.
+  ArrayBuilder(const BuilderOptions &options);
+
+  /// @brief Copy the current snapshot into the BuffersContainer and
+  /// return a Form as a std::string (JSON).
+  const std::string to_buffers(BuffersContainer &container,
+                               int64_t &form_key_id) const;
```

Comment:
Alignment of arguments is good, but black does this better by not having the initial alignment depend on the length of the name of the function. If we rename this function to something like `to_buffers_xyz`, then all of the arguments would get moved and it would be a bigger diff.

If there's any configuration that would change this to

```suggestion
  const std::string to_buffers(
    BuffersContainer &container,
    int64_t &form_key_id,
  ) const;
```

(i.e. black's style), then it would be _better_, but not _necessary_. Note that [trailing commas are allowed in C++ as of C++11](https://stackoverflow.com/a/6372686/1623645).

Reply:
I did not find a simple option for that - unless we go for a `Custom` style.

I kind of like this additional option aligning parameter names, but it will only apply if the line is longer than a limit.
```diff
-AlignConsecutiveDeclarations: None
+AlignConsecutiveDeclarations: true
```
```c++ 
  /// @brief Copy the current snapshot into the BuffersContainer and
  /// return a Form as a std::string (JSON).
  const std::string to_buffers(BuffersContainer &container,
                               int64_t          &form_key_id) const;
```
also it's not very consistent:
```c++
  void concatenate_to(
      PRIMITIVE *to_ptr,
      size_t offset) const noexcept {
    memcpy(to_ptr + offset,
           reinterpret_cast<void *>(
               ptr_.get()),
           length_ * sizeof(PRIMITIVE));
    if (next_) {
      next_->concatenate_to(
          to_ptr, offset + length_);
    }
  }

  /// @brief Fills (one panel) GrowableBuffer<TO_PRIMITIVE> with the
  /// elements of (possibly multi-panels) GrowableBuffer<PRIMITIVE>.
  ///
  /// Changes the data type from `PRIMITIVE` to `TO_PRIMITIVE`/
  template <typename TO_PRIMITIVE>
  void copy_as(TO_PRIMITIVE *to_ptr,
               size_t        offset) {
    for (size_t i = 0; i < length_;
         i++) {
      to_ptr[offset++] =
          static_cast<TO_PRIMITIVE>(
              ptr_.get()[i]);
    }
    if (next_) {
      next_->copy_as(to_ptr, offset);
    }
  }

```

Reply:
Hmm, it looks like it does it in the `for` loop, which is good, but I think it's too aggressive here - the line length isn't too long yet.

RE the argument parameters, maybe we want to set `InsertTrailingCommas` and `AllowAllParametersOfDeclarationOnNextLine`?

Sympathies with you @ianna, I've done this before and it's painful :(